### PR TITLE
wg_engine: introduced composition ability

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
[issues 1479: Masking](#1479)

Supported composition methods:

    AlphaMask
    InvAlphaMask
    LumaMask
    InvLumaMask
    AddMask
    SubtractMask
    IntersectMask
    DifferenceMask

Usage example:

    //Solid Rectangle
    auto shape = tvg::Shape::gen();
    shape->appendRect(0, 0, 400, 400);
    shape->fill(0, 0, 255);

    //Mask
    auto mask = tvg::Shape::gen();
    mask->appendCircle(200, 200, 125, 125);
    mask->fill(255, 255, 255);        //AlphaMask RGB channels are unused.

    //Nested Mask
    auto nMask = tvg::Shape::gen();
    nMask->appendCircle(220, 220, 125, 125);
    nMask->fill(255, 255, 255);       //AlphaMask RGB channels are unused.

    mask->composite(std::move(nMask), tvg::CompositeMethod::AlphaMask);
    shape->composite(std::move(mask), tvg::CompositeMethod::AlphaMask);
    canvas->push(std::move(shape));

    //Star
    auto star = tvg::Shape::gen();
    star->fill(80, 80, 80);
    star->moveTo(599, 34);
    star->lineTo(653, 143);
    star->lineTo(774, 160);
    star->lineTo(687, 244);
    star->lineTo(707, 365);
    star->lineTo(599, 309);
    star->lineTo(497, 365);
    star->lineTo(512, 245);
    star->lineTo(426, 161);
    star->lineTo(546, 143);
    star->close();
    star->strokeWidth(30);
    star->strokeJoin(tvg::StrokeJoin::Miter);
    star->strokeFill(255, 255, 255);

    //Mask3
    auto mask3 = tvg::Shape::gen();
    mask3->appendCircle(600, 200, 125, 125);
    mask3->fill(255, 255, 255);       //AlphaMask RGB channels are unused.
    mask3->opacity(200);
    star->composite(std::move(mask3), tvg::CompositeMethod::AlphaMask);
    if (canvas->push(std::move(star)) != tvg::Result::Success) return;

### CODE REVIEW

File: src/renderer/wg_engine/tvgWgPipelines.cpp [hunk 1]
```diff
@@ -223,33 +223,171 @@ void WgPipelineImage::initialize(WGPUDevice device)
...
+    solid.release();
+    fillStroke.release();
+    fillShape.release();
+}
+
+
+WgPipelineComposition* WgPipelines::getCompositionPipeline(CompositeMethod method)
+{
+    switch (method) {
+        case CompositeMethod::ClipPath:
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgPipelines.cpp:381]: Alpha blending does not take into account the ClipPath. it's more like a scissor behavior.

File: src/renderer/wg_engine/tvgWgPipelines.cpp [hunk 2]
```diff
@@ -223,33 +223,171 @@ void WgPipelineImage::initialize(WGPUDevice device)
...
+    fillStroke.initialize(device);
+    solid.initialize(device);
+    linear.initialize(device);
+    radial.initialize(device);
+    image.initialize(device);
+    blit.initialize(device);
+    blitColor.initialize(device);
+    // composition pipelines
+    compAlphaMask.initialize(device, cShaderSource_PipelineCompAlphaMask);
+    compInvAlphaMask.initialize(device, cShaderSource_PipelineCompInvAlphaMask);
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgPipelines.cpp:341]: Except for alpha masking, the other features are quite optional. They are not frequently used in many animations. That would be better if we can initialize them in a deffered way.

File: src/renderer/wg_engine/tvgWgRenderer.cpp [hunk 3]
```diff
@@ -273,26 +275,99 @@ bool WgRenderer::target(void* window, uint32_t w, uint32_t h)
...
     assert(mSwapChain);
 
-    mRenderTarget.initialize(mContext, mPipelines, w, h);
+    mRenderTargetWnd.initialize(mContext, mPipelines, w, h);
+    mRenderTargetRoot.initialize(mContext, mPipelines, w, h);
     return true;
 }
 
 
 Compositor* WgRenderer::target(TVG_UNUSED const RenderRegion& region, TVG_UNUSED ColorSpace cs)
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgRenderer.cpp:284]: Reffering composition region would help to reduce the composition area significantly.

**Review by hermet — APPROVED** [review]
@SergeyLebedkin LGTM for now, just in case, please double-check my comments for future development.

File: src/renderer/wg_engine/tvgWgPipelines.cpp (hunk 2 above)
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgPipelines.cpp:341]: > Except for alpha masking, the other features are quite optional. They are not frequently used in many animations. That would be better if we can initialize them in a deffered way.

Good point. Will use RAII for it

**Pull request merged**